// @deck.gl/arcgis/dist/es5/deck-layer-view-2d.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createDeckLayerView2D;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _commons = require("./commons");

function createDeckLayerView2D(BaseLayerViewGL2D) {
  return BaseLayerViewGL2D.createSubclass({
    properties: {
      deckgl: {},
      deckFbo: {},
      model: {},
      buffer: {}
    },
    attach: function attach() {
      var _this = this;

      var gl = this.context;

      _commons.initializeResources.call(this, gl);

      this.layer.deck.on('change', function (props) {
        return _this.deckInstance.setProps(props);
      });
      this.deckInstance.setProps(this.layer.deck.toJSON());
    },
    redraw: function redraw() {
      this.requestRender();
    },
    detach: function detach() {
      _commons.finalizeResources.call(this);
    },
    render: function render(renderParameters) {
      var _this$view$state$size = (0, _slicedToArray2["default"])(this.view.state.size, 2),
          width = _this$view$state$size[0],
          height = _this$view$state$size[1];

      var state = renderParameters.state;

      _commons.render.call(this, {
        gl: renderParameters.context,
        width: width,
        height: height,
        viewState: {
          latitude: this.view.center.latitude,
          longitude: this.view.center.longitude,
          zoom: this.view.featuresTilingScheme.scaleToLevel(state.scale),
          bearing: -state.rotation,
          pitch: 0
        }
      });
    }
  });
}
