// @deck.gl/arcgis/dist/es5/deck-renderer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createDeckRenderer;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _commons = require("./commons");

function createDeckRenderer(DeckProps, externalRenderers) {
  var DeckRenderer = function () {
    function DeckRenderer(view, props) {
      (0, _classCallCheck2["default"])(this, DeckRenderer);
      this.view = view;
      this.deck = new DeckProps(props);
    }

    (0, _createClass2["default"])(DeckRenderer, [{
      key: "setup",
      value: function setup(context) {
        var _this = this;

        var gl = context.gl;

        _commons.initializeResources.call(this, gl);

        this.deck.on('change', function (props) {
          return _this.deckInstance.setProps(props);
        });
        this.deckInstance.setProps(this.deck.toJSON());
      }
    }, {
      key: "dispose",
      value: function dispose() {
        _commons.finalizeResources.call(this);
      }
    }, {
      key: "redraw",
      value: function redraw() {
        externalRenderers.requestRender(this.view);
      }
    }, {
      key: "render",
      value: function render(context) {
        var _this$view$size = (0, _slicedToArray2["default"])(this.view.size, 2),
            width = _this$view$size[0],
            height = _this$view$size[1];

        _commons.render.call(this, {
          gl: context.gl,
          width: width,
          height: height,
          viewState: {
            latitude: this.view.center.latitude,
            longitude: this.view.center.longitude,
            zoom: this.view.zoom,
            bearing: this.view.camera.heading,
            pitch: this.view.camera.tilt
          }
        });
      }
    }]);
    return DeckRenderer;
  }();

  return DeckRenderer;
}
