// @deck.gl/arcgis/dist/esm/deck-props.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
var properties = {
  layers: {},
  layerFilter: {},
  parameters: {},
  effects: {},
  pickingRadius: {},
  onBeforeRender: {},
  onAfterRender: {},
  onClick: {},
  onHover: {},
  onDragStart: {},
  onDrag: {},
  onDragEnd: {},
  onError: {},
  debug: {},
  drawPickingColors: {}
};
export default function loadDeckProps(Accessor) {
  var DeckProps = Accessor.createSubclass({
    properties: properties,
    constructor: function constructor() {
      var _this = this;

      this._callbacks = {};
      this.watch(Object.keys(properties), function (newValue, oldValue, propName) {
        _this.emit('change', _defineProperty({}, propName, newValue));
      });
    },
    on: function on(eventName, cb) {
      this._callbacks[eventName] = this._callbacks[eventName] || [];

      this._callbacks[eventName].push(cb);
    },
    emit: function emit(eventName, details) {
      var callbacks = this._callbacks[eventName];

      if (callbacks) {
        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = callbacks[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var cb = _step.value;
            cb(details);
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator["return"] != null) {
              _iterator["return"]();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }
      }
    },
    toJSON: function toJSON() {
      var result = {};

      for (var key in this.properties) {
        if (this[key] !== undefined) {
          result[key] = this[key];
        }
      }

      return result;
    }
  });
  return DeckProps;
}
