// @deck.gl/arcgis/dist/esm/deck-layer-view-2d.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { initializeResources, render as _render, finalizeResources } from './commons';
export default function createDeckLayerView2D(BaseLayerViewGL2D) {
  return BaseLayerViewGL2D.createSubclass({
    properties: {
      deckgl: {},
      deckFbo: {},
      model: {},
      buffer: {}
    },
    attach: function attach() {
      var _this = this;

      var gl = this.context;
      initializeResources.call(this, gl);
      this.layer.deck.on('change', function (props) {
        return _this.deckInstance.setProps(props);
      });
      this.deckInstance.setProps(this.layer.deck.toJSON());
    },
    redraw: function redraw() {
      this.requestRender();
    },
    detach: function detach() {
      finalizeResources.call(this);
    },
    render: function render(renderParameters) {
      var _this$view$state$size = _slicedToArray(this.view.state.size, 2),
          width = _this$view$state$size[0],
          height = _this$view$state$size[1];

      var state = renderParameters.state;

      _render.call(this, {
        gl: renderParameters.context,
        width: width,
        height: height,
        viewState: {
          latitude: this.view.center.latitude,
          longitude: this.view.center.longitude,
          zoom: this.view.featuresTilingScheme.scaleToLevel(state.scale),
          bearing: -state.rotation,
          pitch: 0
        }
      });
    }
  });
}
