// @deck.gl/arcgis/dist/esm/index.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import createDeckProps from './deck-props';
import createDeckLayer from './deck-layer';
import createDeckLayerView2D from './deck-layer-view-2d';
import createDeckRenderer from './deck-renderer';
import { loadModules as esriLoaderLoadModules } from 'esri-loader';
var arcGIS = null;
export function loadArcGISModules(_x, _x2) {
  return _loadArcGISModules.apply(this, arguments);
}

function _loadArcGISModules() {
  _loadArcGISModules = _asyncToGenerator(_regeneratorRuntime.mark(function _callee(modules, loadScriptOptions) {
    var namespace;
    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            namespace = Array.isArray(modules) ? null : modules;
            _context.next = 3;
            return loadArcGISModule(namespace, loadScriptOptions);

          case 3:
            if (!Array.isArray(modules)) {
              _context.next = 5;
              break;
            }

            return _context.abrupt("return", esriLoaderLoadModules(modules, loadScriptOptions).then(function (array) {
              return _objectSpread({}, arcGIS, {
                modules: array
              });
            }));

          case 5:
            return _context.abrupt("return", arcGIS);

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _loadArcGISModules.apply(this, arguments);
}

function loadArcGISModule(esri, loadScriptOptions) {
  if (arcGIS) {
    return arcGIS;
  }

  if (esri) {
    var Layer = esri.layers.Layer;
    var Accessor = esri.core.Accessor;
    var BaseLayerViewGL2D = esri.views['2d'].layers.BaseLayerViewGL2D;
    var externalRenderers = esri.views['3d'].externalRenderers;
    return initialize(Layer, Accessor, BaseLayerViewGL2D, externalRenderers);
  }

  return esriLoaderLoadModules(['esri/layers/Layer', 'esri/core/Accessor', 'esri/views/2d/layers/BaseLayerViewGL2D', 'esri/views/3d/externalRenderers'], loadScriptOptions).then(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 4),
        Layer = _ref2[0],
        Accessor = _ref2[1],
        BaseLayerViewGL2D = _ref2[2],
        externalRenderers = _ref2[3];

    return initialize(Layer, Accessor, BaseLayerViewGL2D, externalRenderers);
  });
}

function initialize(Layer, Accessor, BaseLayerViewGL2D, externalRenderers) {
  var DeckProps = createDeckProps(Accessor);
  var DeckLayerView2D = createDeckLayerView2D(BaseLayerViewGL2D);
  var DeckLayer = createDeckLayer(DeckProps, Layer, DeckLayerView2D);
  var DeckRenderer = createDeckRenderer(DeckProps, externalRenderers);
  arcGIS = {
    DeckLayer: DeckLayer,
    DeckRenderer: DeckRenderer
  };
  return arcGIS;
}
